### ISSUE TITLE
Paths on Windows are invalid

### ISSUE THREAD

**Issue by 71** [Canop/broot#78]
**Describe the bug**
On Windows, the extended path with UNC prefix are used to canonicalize paths (e.g in the root directory at the top, or in output of commands). This causes issues

**To Reproduce**
Using the following Powershell Core function, try to `:cd` into a directory.

```pwsh
function br
{
    $TempFile = New-TemporaryFile
    broot.exe --outcmd $TempFile $Args

    if (!$?)
    {
        Remove-Item -Force $TempFile
        return $LASTEXITCODE
    }

    $CommandToExecute = Get-Content $TempFile
    Remove-Item -Force $TempFile
    Invoke-Expression $CommandToExecute
}
```

An error will be shown, indicating that the path `\\?\A:\Foo` does not exist (even though `A:\Foo` does exist).

**Expected behavior**
Normal (usable) links with no UNC prefix should be used, since most commands do not support those prefixes.

**Configuration**
- OS: Windows
- Version: 10.0.18362

**Additional context**
As first reported [here](https://github.com/Canop/broot/issues/50#issuecomment-571059434), the problem is that Broot [calls](https://github.com/Canop/broot/blob/d327ae6cb0fd00b66f4ca33664b904684269e9e0/src/cli.rs#L151) [`std::fs::canonicalize`](https://doc.rust-lang.org/std/fs/fn.canonicalize.html), which returns a path like `\\?\C:\Users\71`, instead of `C:\Users\71`. Therefore using `:cd` inside Broot will write a command `cd '\\?\A:\foo'`, which will confuse PowerShell (and most commands in general).

A workaround is to execute `$Command.Replace('\\?\', '')` instead of `$Command`.

In the end, the script I ended up using is:

```pwsh
function br
{
    $TempFile = New-TemporaryFile
    broot.exe --outcmd $TempFile $Args

    if (!$?)
    {
        Remove-Item -Force $TempFile
        return $LASTEXITCODE
    }

    $CommandToExecute = Get-Content $TempFile
    Remove-Item -Force $TempFile
    Invoke-Expression $CommandToExecute.Replace('\\?\', '')
}
```

**However**, it looks like this **does not work** in all cases. [This crate](https://gitlab.com/kornelski/dunce#dunce-de-unc) attempts to make the transformation in a smart way, but I wonder what should happen when it judges that the UNC prefix cannot be stripped.

**Comment by Canop** [comment]
I can't test on windows so I'll need your help for that (you already wrote a beautiful bug report).

Do you think we could just solve the problem by issuing our own canonicalization at broot launch?

Something like this:

```rust
#[cfg(not(windows))]
fn canonicalize_root(root: &Path) -> io::Result<PathBuf> {
    root.canonicalize()
}

#[cfg(windows)]
fn canonicalize_root(root: &Path) -> io::Result<PathBuf> {
    Ok(
        if root.is_relative() {
            env::current_dir()?.join(root)
        } else {
            root.to_path_buf()
        }
    )
}
```

**Comment by Canop** [comment]
note: the referenced commit isn't on master right now (it's in a "no-unce" branch)

**Comment by stinos** [comment]
I'd be happy to help testing/adding docs. But I can't get started: I defined the `br` function but still when launching it all I get is
```
Broot should be launched using a shell function (see https://github.com/Canop/broot for explanations).
The function is either missing, old or badly installed.
Can I install it now? [Y n]

Removing C:\Users\<user>\AppData\Roaming\dystroy\broot\data\launcher\bash\1.
Writing br shell function in C:\Users\<user>\AppData\Roaming\dystroy\broot\data\launcher\bash\1.
Removing C:\Users\<user>\AppData\Roaming\dystroy\broot\config\launcher\bash\br.
Creating link from C:\Users\<user>\AppData\Roaming\dystroy\broot\config\launcher\bash\br to C:\Users\<user>\AppData\Roaming\dystroy\broot\data\launcher\bash\1.

I found neither .bashrc nor .zshrc.
If you're using bash or zsh, then installation isn't complete:
the br function initialization script won't be sourced unless you source it yourself.

You cannot call a method on a null-valued expression.
At C:\Users\<user>\Documents\WindowsPowerShell\PowerShell_profile_curhost.ps1:31 char:5
+     Invoke-Expression $CommandToExecute.Replace('\\?\', '')
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
```
(that last error is from the br function, I assume broot didn't return non-zero exit code making the function continue)

Any hints?

**Comment by Canop** [comment]
@stinos I just published a new version fixing the 'n' answer being ignored (a problem I discovered yesterday).

**Comment by stinos** [comment]
Can confirm that works. Just get a black screen upon launching though, @71 did you do anything special to get this working? If I get some time I'll install a Rust environment for building from source.

**Comment by lubieowoce** [comment]
@71 on my system your function crashes if broot doesn't emit a cd command, i.e the file is empty. i'm pretty new to PS, but this version seems to work for me:

```powershell
function br {
    $outf = new-temporaryfile
    broot.exe --outcmd $outf $args

    if (!$?) {
        remove-item -force $outf
        return $lastexitcode
    }

    $command = get-content $outf
    if ($command) {
        $command = $command.replace("\\?\", "", 1)
        invoke-expression $command
    }
    remove-item -force $outf
}
```

**Comment by 71** [comment]
@stinos I did have to manually create the file `%APPDATA%\dystroy\broot\config\launcher\installed-v1`.

@lubieowoce Arh I forgot PowerShell likes to implicitly convert everything... Good catch!

**Comment by stinos** [comment]
>  I did have to manually create the file ...

@71 I already got the file `refused` there since Canop fixed the issue mentioned above, so that isn't the problem. I think it has to do with rendering/esacpe codes (it does react to Esc key for instance i.e. it's like running headless). If I launch broot in in plain cmd or PS I just get a blank screen. If I launch it in cmd running under Conemu though it printes a ton of escape codes e.g. here's an excerpt:

```
←[68;1H←[K←[68;1H←[48;5;235m←[38;5;252m ←[0m←[48;5;235m←[38;5;252mHit ←[0m←[48;5;235m←[38;5;178mesc←[0m←[48;5;235m←[38;5;252m to go back, ←[0m←[48;5;235m←[38;5;178menter←[0m←[48;5;235m←[38;5;252m to go up, ←[0m←[48;5;235m←[38;5;178m?←[0m←[48;5;235m←[38;5;252m for help, or a few letters to search←[0m←[48;5;235m←[38;5;252m
```

@Canop this is going off topic, make a new issue for this or reopen e.g. the 'Windws compatibility' one?

**Comment by Canop** [comment]
I don't want to reopen a too wide issue. I'd like clear separate issues, with reproduction steps and description of the problems, so that they can be addressed.

**Comment by lubieowoce** [comment]
how about a separate issue for powershell support/integration? that does seem kind of off-topic here, and it'd be easier to find for people looking for a suitable `br` function

**Comment by stinos** [comment]
> An error will be shown, indicating that the path \\?\A:\Foo does not exist (even though A:\Foo does exist).

Which PS version? I cannot reproduce this, i.e. running Windows 10 version 1803 with PSVersion 5.1.17134.590 the command `cd \\?\A:\Foo` works, though leads to cd'ing to ` Microsoft.PowerShell.Core\FileSystem::\\?\A:\Foo`. Don't know when this changed exactly but since a while Windows has extended support for paths like this, e.g. when broot runs `$EDITOR \\?\A:\myfile.txt` that also works for me.

**Comment by 71** [comment]
@stinos I'm on PS Core (v6.2.2, Windows 10 version 1903), which is a little less Windows-focused than regular PowerShell. `cd \\?\A:\Foo` doesn't work, `cd A:\Foo` does. Maybe there's a setting somewhere else...

**Comment by Canop** [comment]
TLDR: obsolete issue that I should have closed long ago

Paths are no longer canonicalized on Windows, so the \\?\ UNC prefix doesn't appear anymore (see #809 and #812 for the last fixes). The PowerShell br function is now installed by broot itself. Closing this old one; please open a new issue with reproduction steps for any remaining Windows path problem.
